### PR TITLE
Multi-column mapping API for normalizer estimators.

### LINKED ISSUE #3171: Multi-column mapping API for normalizer estimators.
We need an API that lets the user normalize multiple columns at once for ease and more importantly performance benefits.

### PR DESCRIPTION
**Pull request by codemzs**
fixes #3171

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/Normalizer.cs [hunk 1]
```diff
@@ -56,8 +56,8 @@ public static void Example()
 
             // Composing a different pipeline if we wanted to normalize more than one column at a time. 
             // Using log scale as the normalization mode. 
-            var multiColPipeline = ml.Transforms.NormalizeMinMax("LogInduced", "Induced")
-                .Append(ml.Transforms.NormalizeMinMax("LogSpontaneous", "Spontaneous"));
+            var multiColPipeline = ml.Transforms.NormalizeLogMeanVariance(new[] { new InputOutputColumnPair("LogInduced", "Induced"), new InputOutputColumnPair("LogSpontaneous", "Spontaneous") });
```
Comment (eerhardt) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Normalizer.cs]: Why the change from `NormalizeMinMax` to `NormalizeLogMeanVariance`?

Comment (codemzs) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Normalizer.cs]: Hey Eric, it was initially LogMeanVariance but last change to this file accidentally changed to MinMax. I also verified the outputs are same as the comments. You can see output column names have "Log"

---
In reply to: [271416903](https://github.com/dotnet/machinelearning/pull/3172#discussion_r271416903) [](ancestors = 271416903)

File: src/Microsoft.ML.Transforms/NormalizerCatalog.cs [hunk 2]
```diff
@@ -1,4 +1,5 @@
-﻿using Microsoft.ML.Data;
+﻿using System.Linq;
```
Comment (eerhardt) [inline src/Microsoft.ML.Transforms/NormalizerCatalog.cs]: While you are in here - can you add a copyright to this file?

Comment (codemzs) [reply src/Microsoft.ML.Transforms/NormalizerCatalog.cs]: totally, good catch.

---
In reply to: [271425802](https://github.com/dotnet/machinelearning/pull/3172#discussion_r271425802) [](ancestors = 271425802)

**Comment by eerhardt** [comment]
Can we add some tests for the new APIs?

File: test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs [hunk 3]
```diff
@@ -207,6 +208,118 @@ public void NormalizerParameters()
...
+                    new TextLoader.Column("double4", DataKind.Double, new[]{new TextLoader.Range(1, 4) }),
+                    new TextLoader.Column("int1", DataKind.Int32, 0),
+                    new TextLoader.Column("float0", DataKind.Single, new[]{ new TextLoader.Range { Min = 1, VariableEnd = true } })
+                },
+                HasHeader = true
+            }, new MultiFileSource(dataPath));
+            var context = new MLContext(seed: 0);
+            var est = context.Transforms.NormalizeMinMax(
+                new[] { new InputOutputColumnPair("float1"), new InputOutputColumnPair("float4"),
+                    new InputOutputColumnPair("double1"), new InputOutputColumnPair("double4"), }).Append(
```
Comment (eerhardt) [inline test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs]: (nit) I think this reads better if you put the `Append` on a new line, and not at the end of the previous line.

```C#
            var est = context.Transforms.NormalizeMinMax(
                new[] { new InputOutputColumnPair("float1"), new InputOutputColumnPair("float4"),
                    new InputOutputColumnPair("double1"), new InputOutputColumnPair("double4"), })
                .Append(context.Transforms.NormalizeBinning(
                        new[] {new InputOutputColumnPair("float1bin", "float1"), new InputOutputColumnPair("float4bin", "float4"),
                            new InputOutputColumnPair("double1bin", "double1"), new InputOutputColumnPair("double4bin", "double4")}))
                .Append(context.Transforms.NormalizeMeanVariance(
                        new[] {new InputOutputColumnPair("float1mv", "float1"), new InputOutputColumnPair("float4mv", "float4"),
                            new InputOutputColumnPair("double1mv", "double1"), new InputOutputColumnPair("double4mv", "double4")}))
                .Append(context.Transforms.NormalizeLogMeanVariance(
                        new[] {new InputOutputColumnPair("float1lmv", "float1"), new InputOutputColumnPair("float4lmv", "float4"),
                            new InputOutputColumnPair("double1lmv", "double1"), new InputOutputColumnPair("double4lmv", "double4")}));
```

File: test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs [hunk 4]
```diff
@@ -207,6 +208,118 @@ public void NormalizerParameters()
...
+                            new InputOutputColumnPair("double1mv", "double1"), new InputOutputColumnPair("double4mv", "double4")})).Append(
+                    context.Transforms.NormalizeLogMeanVariance(
+                        new[] {new InputOutputColumnPair("float1lmv", "float1"), new InputOutputColumnPair("float4lmv", "float4"),
+                            new InputOutputColumnPair("double1lmv", "double1"), new InputOutputColumnPair("double4lmv", "double4")}));
+
+            var data = loader.Load(dataPath);
+
+            var transformer = est.Fit(data);
+            var transformers = transformer.ToImmutableArray();
+            var floatAffineData = (transformers[0] as NormalizingTransformer).Columns[0].ModelParameters as NormalizingTransformer.AffineNormalizerModelParameters<float>;
```
Comment (eerhardt) [inline test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs]: These should be casts and not `as`. You don't expect these to fail, so getting an `InvalidCastException` is better than getting a `NullReferenceException` if they do happen to fail.

File: test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs [hunk 5]
```diff
@@ -207,6 +208,118 @@ public void NormalizerParameters()
...
+                    new InputOutputColumnPair("double1"), new InputOutputColumnPair("double4"), }).Append(
+                    context.Transforms.NormalizeBinning(
+                        new[] {new InputOutputColumnPair("float1bin", "float1"), new InputOutputColumnPair("float4bin", "float4"),
+                            new InputOutputColumnPair("double1bin", "double1"), new InputOutputColumnPair("double4bin", "double4")})).Append(
+                    context.Transforms.NormalizeMeanVariance(
+                        new[] {new InputOutputColumnPair("float1mv", "float1"), new InputOutputColumnPair("float4mv", "float4"),
+                            new InputOutputColumnPair("double1mv", "double1"), new InputOutputColumnPair("double4mv", "double4")})).Append(
+                    context.Transforms.NormalizeLogMeanVariance(
+                        new[] {new InputOutputColumnPair("float1lmv", "float1"), new InputOutputColumnPair("float4lmv", "float4"),
+                            new InputOutputColumnPair("double1lmv", "double1"), new InputOutputColumnPair("double4lmv", "double4")}));
```
Comment (eerhardt) [inline test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs]: How about `NormalizeSupervisedBinning`? Do we have a test for that?

Comment (codemzs) [reply test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs]: Just added within this test.

---
In reply to: [271496017](https://github.com/dotnet/machinelearning/pull/3172#discussion_r271496017) [](ancestors = 271496017)

File: test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs [hunk 6]
```diff
@@ -207,6 +208,144 @@ public void NormalizerParameters()
...
+            Assert.Equal(4, doubleAffineDataVec.Scale.Length);
+            Assert.Empty(doubleAffineDataVec.Offset);
+
+            var floatBinData = ((NormalizingTransformer)transformers[1]).Columns[0].ModelParameters as NormalizingTransformer.BinNormalizerModelParameters<float>;
+            Assert.True(35 == floatBinData.UpperBounds.Length);
+            Assert.True(34 == floatBinData.Density);
+            Assert.True(0 == floatBinData.Offset);
+
+            var floatBinDataVec = ((NormalizingTransformer)transformers[1]).Columns[1].ModelParameters as NormalizingTransformer.BinNormalizerModelParameters<ImmutableArray<float>>;
+            Assert.True(4 == floatBinDataVec.UpperBounds.Length);
```
Comment (wschin) [inline test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs]: Could you please check at least the first and the last values in `UpperBounds` and other array-like fields?

File: test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs [hunk 7]
```diff
@@ -207,6 +208,144 @@ public void NormalizerParameters()
...
+                                    new InputOutputColumnPair("double1lmv", "double1"), new InputOutputColumnPair("double4lmv", "double4")}))
+                    .Append(context.Transforms.NormalizeSupervisedBinning(
+                                new[] {new InputOutputColumnPair("float1nsb", "float1"), new InputOutputColumnPair("float4nsb", "float4"),
+                                    new InputOutputColumnPair("double1nsb", "double1"), new InputOutputColumnPair("double4nsb", "double4")}));
+
+            var data = loader.Load(dataPath);
+
+            var transformer = est.Fit(data);
+            var transformers = transformer.ToImmutableArray();
+            var floatAffineData = ((NormalizingTransformer)transformers[0]).Columns[0].ModelParameters as NormalizingTransformer.AffineNormalizerModelParameters<float>;
```
Comment (wschin) [inline test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs]:
```suggestion
            var floatAffineModel = ((NormalizingTransformer)transformers[0]).Columns[0].ModelParameters as NormalizingTransformer.AffineNormalizerModelParameters<float>;
```
This doesn't look like `data`.

Comment (codemzs) [reply test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs]: yeah its just a test and modified version of the above test but ok.

---
In reply to: [271538809](https://github.com/dotnet/machinelearning/pull/3172#discussion_r271538809) [](ancestors = 271538809)
